src/MQTTClient.php

```diff
@@ -143,29 +155,114 @@ public function connect(
-            $this->socket     = stream_socket_client($connectionString, $errorCode, $errorMessage, 60, STREAM_CLIENT_CONNECT, $socketContext);
-        } else {
-            $connectionString = 'tcp://' . $this->getHost() . ':' . $this->getPort();
-            $this->socket     = stream_socket_client($connectionString, $errorCode, $errorMessage, 60, STREAM_CLIENT_CONNECT);
+        $useTls = ($this->settings->shouldUseTls() || $this->hasCertificateAuthorityFile());
+        $contextOptions = [];
+
+        if ($useTls) {
+            $contextOptions['ssl'] = [];
+
+            if ($this->settings->shouldTlsVerifyPeer() || $this->hasCertificateAuthorityFile()) {
+                // As it does not make any sense to specify a certificate file without
+                // enabling peer verification, we automatically enable it in that case
+                $contextOptions['ssl']['verify_peer'] = true;
+            }
+            else {
+                // You get a warning output for free, on every connection
+                $this->logger->warning('Enabled TLS encryption without peer verification - DANGEROUS SECURITY ISSUE');
+                $contextOptions['ssl']['verify_peer'] = false;
+            }
+
+            $contextOptions['ssl']['verify_peer_name'] = $this->settings->shouldTlsVerifyPeerName();
+
+            if ($this->hasCertificateAuthorityFile()) {
+                $this->logger->info(sprintf('Using certificate authority file [%s] to verify peer name.', $this->caFile));
+                $contextOptions['ssl']['cafile'] = $this->getCertificateAuthorityFile();
+            }
+        }
+
+        $connectionString = 'tcp://' . $this->getHost() . ':' . $this->getPort();
+        $socketContext = stream_context_create($contextOptions);
+        $socket = @stream_socket_client(
+            $connectionString,
+            $errorCode,
+            $errorMessage,
+            $this->settings->getSocketTimeout(),
+            STREAM_CLIENT_CONNECT,
+            $socketContext
+        );
+
+        if ($socket === false) {
+            $this->logger->error(sprintf('Establishing a connection with the MQTT broker using connection string [%s] failed: %s (code %d).', $connectionString, $errorMessage, $errorCode));
+            // errorCode is a POSIX socket error (0-255), so we map it to the 1000-1255 range
+            // FIXME: the above is not true for TLS errors...unfortunately
+            throw new ConnectingToBrokerFailedException(
+                self::EXCEPTION_CONNECTION_SOCKET_ERROR + $errorCode,
+                "Socket error: " . $errorMessage
+            );
         }
 
-        if ($this->socket === false) {
-            $this->logger->error(sprintf('Establishing a connection with the MQTT broker using connection string [%s] failed.', $connectionString));
-            throw new ConnectingToBrokerFailedException($errorCode, $errorMessage);
+        if ($useTls) {
+            error_clear_last();
```

Comment:
Is this really a good idea?

Reply:
You mean to clear the error? You have no other choice, as you can't guarantee the next call will *actually* generate a PHP error on failure, clearing the error guarantees if you get one it's from there.

Otherwise it could be something happened in a totally unrelated place, like a @fopen() from another library.

Reply:
I see. Feels a bit dirty to me, but I guess that's just how PHP has to be made working.